LanguageFeatures/Horizontal-inference/bounds_t01.dart

```diff
@@ -0,0 +1,31 @@
+// Copyright (c) 2022, the Dart project authors.  Please see the AUTHORS file
+// for details. All rights reserved. Use of this source code is governed by a
+// BSD-style license that can be found in the LICENSE file.
+
+/// @assertion Horizontal inference allows certain arguments of an invocation
+/// (specifically function literals) to be type inferred in a context derived
+/// from the type of other arguments.
+///
+/// @description Checks horizontal inference when type bounds specified
+/// @author sgrekhov22@gmail.com
+
+import "../../Utils/expect.dart";
+
+class C1 {}
+class C2 extends C1 {}
+class C3 extends C2 {
+  void c3() {}
+}
+
+void f1<T extends C1>(void Function(T) a, T b) {
+  Expect.equals(T, C3);
```

Comment:
@eernstg did you mean 
```dart 
Expect.isTrue(a is void Function(C3));
``` 
here? Because
```dart
void f1<T extends C1>(void Function(T) a, T b) {
  Expect.isTrue(a is void Function(T));
```
will always be true

Reply:
That's true, `a is void Function(T)` should be guarantee by soundness already, so we wouldn't expect any type inference bugs to be able to violate that.

What we actually need is `Expect.equals(typeOf<void Function(T)>(), a.runtimeType)`, because that would fit the expected behavior when the call site enables transfer of information from the actual argument for `b` to a function literal which is the actual argument for `a`.

(And I guess we're relying on `.runtimeType == someFunctionType` in so many ways that this extra case makes no difference.)